WIP: Add PSBT derivation info for remote signer

aakselrod opened the pull request:
## Change Description
Description of change / link to associated issue.

## Steps to Test
Steps for reviewers to follow to test the change.

## Pull Request Checklist
### Testing
- [ ] Your PR passes all CI checks.
- [ ] Tests covering the positive and negative (error paths) are included.
- [ ] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [ ] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [ ] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [ ] Any new logging statements use an appropriate subsystem and logging level.
- [ ]  [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our  [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

guggero commented on input/psbt.go:
```diff
@@ -0,0 +1,144 @@
+
+const (
+	// TODO(aakselrod): should this be encoded as a uint/varint instead?
+	// That would save space on the wire.
```
Yes, I think those keys (`<keytype>` in the spec) are supposed to be "A [compact size](https://en.bitcoin.it/wiki/Protocol_documentation#Variable_length_integer) unsigned integer representing the type.". We have introduced lnd/btcwallet specific ones before, see https://github.com/lightningnetwork/lnd/blob/master/lnwallet/btcwallet/psbt.go#L26.

aakselrod replied:
I've done this, starting at `0x60` to not overlap with `0x51` for single tweaks, even though that's currently used per-input rather than per-output.

guggero commented on input/psbt.go:
```diff
@@ -0,0 +1,144 @@
+
+// LocalDesc returns an UnknownOption for the local descriptor.
+func LocalDesc(desc *keychain.KeyDescriptor) UnknownOption {
+	return wrapUnknownOption(OptLocalDesc, descBytes(desc))
```
This would require a bit more encoding/decoding effort and a bit more space, but would be more "standard" if we used the full BIP32 derivation path here instead, see https://github.com/btcsuite/btcd/blob/master/btcutil/psbt/bip32.go#L65

aakselrod replied:
I'm still working on this, as it requires passing the master key fingerprint and network parameters/coin type through to the PSBT.

guggero commented on input/signdescriptor.go:
```diff
@@ -80,6 +81,13 @@ type SignDescriptor struct {
+	// match exactly the outputs of the transaction that is to be signed,
+	// and should contain any derivation information required by the signer
+	// to authorize the signature in the Unknowns field.
+	POutputs []psbt.POutput
```
Should we mention here that these outputs shouldn't contain any information otherwise expressible in the `SignDescriptor` struct? Because all of the existing fields have an existing encoding path in a `psbt.POutput`. Maybe this could be a slice of a slice of unknowns instead, to make it more clear?

aakselrod replied:
I've changed this to a slice of a slice of unknowns instead, to make it clear.

aakselrod replied:
I've updated the comments as well.

guggero commented on input/signdescriptor.go:
```diff
@@ -307,3 +315,29 @@ func ReadSignDescriptor(r io.Reader, sd *SignDescriptor) error {
+		if packet.Outputs[i].Unknowns != nil &&
+			len(packet.Outputs[i].Unknowns) != 0 {
+
+			return fmt.Errorf("partial output %d already has "+
```
This might already be the case if we encoded the single or double tweak (depending on when/where this would be called).

aakselrod replied:
I've removed the error condition and instead used append, in case there's already data there.

guggero reviewed (commented):
Very excited about this PR and what it will allow the remote signer to validate. Gave it a quick glance even though it's still WIP. But approach looks great to me and it looks like the changes necessary to get this meta information delivered to the remote signer are pretty compact. Nice work!

aakselrod commented:
Thanks for the initial review, @guggero! I've addressed most of the comments, but am still working on the PR.

I think I need to do a bit more work on the chanfunding side, including passing the information required to construct the entire BIP32 derivation. I was working on the assumption that the signer would have the master key fingerprint and coin type already, but that isn't standard for PSBTs. I'll add a preparatory commit for passing that into the assemblers. I think I also need to update one more SignDescriptor that I missed in that package, and add some more testing here.

Roasbeef commented:
Hi @aakselrod, this looks really cool! Can you add some more detail in the PR body which outlines: the motivation, problem at land, this solution, and potential alternatives? This'll help future reviewers get up to speed with the path here if they weren't listening on all the private/offline conversations.

aakselrod commented on lnwallet/test/test_interface.go:
```diff
@@ -3467,3 +3530,339 @@ func runTests(t *testing.T, walletDriver *lnwallet.WalletDriver,
+	// Check for change output. A validating remote signer should look
+	// ahead based on known wallet state, but we cheat here and look it
+	// up in our known addresses. We only do this if the output has no
+	// unknowns.
```
Is this OK, or should we add another unknown with a descriptor here? We'd need to plumb that through as well, through the `ChangeAddr` function when we initialize the wallet assembler.

aakselrod commented on lnwallet/wallet.go:
This is done.

aakselrod commented:
I've done a lot of work in `lnwallet/chanfunding` to ensure that:
- The funding TX contains all the information needed for the remote signer to validate subsequent commitment TXs
- The funding tests check for this info

I think this PR is ready for review. I'm starting work on the `lndsigner` side to start validating signing requests using this new metadata.

lightninglabs-deploy commented:
@guggero: review reminder

guggero commented on input/psbt.go:
```diff
@@ -0,0 +1,327 @@
+	uint32, uint32, keychain.KeyDescriptor, error) {
+
+	// Check if we have a bare derivation or a descriptor with a pubkey.
+	if len(value) != 24 && len(value) != 57 {
```
nit: could use constants for 24 and 57, for example `baseLen` and `baseLen + btcec.PubKeyBytesLenCompressed` to remove some magic numbers.

guggero commented on lnwallet/channel.go:
```diff
@@ -514,6 +515,12 @@ type commitment struct {
 	// chain is being modified, the opposite is true.
 	txn *wire.MsgTx
 
+	// outUnknowns is a slice containing matching derivation informfation to
```
nit: typo in "information".

guggero commented on lnwallet/channel.go:
```diff
@@ -3831,6 +3845,11 @@ func (lc *LightningChannel) SignNextCommitment() (lnwire.Sig, []lnwire.Sig,
 	lc.sigPool.SubmitSignBatch(sigBatch)
 
+	// Ensure that we're passing the correct derivation info to the signer,
+	// and clean up after ourselves.
```
Would be great to add some context here why this clean up is necessary? Is the sign descriptor re-used outside of this method?

guggero commented on lnwallet/channel.go:
```diff
@@ -514,6 +515,12 @@ type commitment struct {
+	// the outputs of the transaction in txn, used to pass to any downstream
+	// signer that can use the information to ensure the channel state
+	// machine is behaving correctly.
+	outUnknowns [][]*psbt.Unknown
```
In another project we created an alias type for `psbt.Unknown`. Since these values aren't really unknown to this code, we could also call the variables differently (e.g. "meta" or "signInfo"). I know that would require a big rename in this PR, so it's really only a light suggestion at this point.

guggero commented on lnwallet/test_utils.go:
```diff
@@ -529,3 +543,209 @@ func ForceStateTransition(chanA, chanB *LightningChannel) error {
+		)
+
+		leaseBytes, hasLease :=
+			unkMap[string(input.PsbtKeyTypeOutputLeaseExpiry)]
```
nit: create local alias for these string converted keys to make the code a bit more easy to read?

guggero commented on lnwallet/test_utils.go:
```diff
@@ -529,3 +543,209 @@ func ForceStateTransition(chanA, chanB *LightningChannel) error {
+			commitPoint, err = btcec.ParsePubKey(commitPointBytes)
+			require.NoError(t, err)
+
+			keyRingOurCommit = DeriveCommitmentKeys(commitPoint,
```
nit: formatting of multi-line function calls here and the rest of the function.

guggero commented on lnwallet/wallet.go:
```diff
@@ -1536,6 +1539,16 @@ func (l *LightningWallet) handleContributionMsg(req *addContributionMsg) {
+		// We also bind the channel configurations to send to the signer
+		// so that a validating remote signer can validate commitment
+		// and second-level HTLC transactions.
+		fundingIntent.ConfigurePsbt(
```
Shouldn't we also call this for the `*chanfunding.PsbtIntent` above? It seems like we have the same information available. And I believe even for the `*chanfunding.ShimIntent`.

Also, this makes me think that we should probably try to add some sort of integration test for this. So we can test that the derivation info is properly attached in different scenarios (e.g. normal channel funding, PSBT based funding, shim based funding).
Not sure how to best approach this? Maybe have the remote signer write each PSBT it is requested to sign to a specific file (would only happen when being compiled with the `integration` build tag), then verify it in the itest?

guggero reviewed (commented):
This looks pretty far along, thanks a lot for all the work! Can only imagine how long it took to identify all these call sites to update.

My main question is around testing, both automated (see inline comment) but also manual testing: Is there a prototype of a policy validation engine available somewhere already that would allow this to be examined/tested more closely?

aakselrod commented:
Thanks for the review, Oliver! I was out for a few days, will update code to address your comments over the next few days.

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

guggero commented on lnwallet/wallet.go:
```diff
@@ -1689,6 +1703,7 @@ func (l *LightningWallet) signCommitTx(pendingReservation *ChannelReservation,
 		signDesc := input.SignDescriptor{
 			WitnessScript: fundingWitnessScript,
 			KeyDesc:       ourKey,
+			OutSignInfo:   theirSignInfo,
```
I think Taproot Channel support should be a priority, given all the benefits we get from them.
So I think changing the `MuSig2Signer` to get an additional parameter is probably the way forward.

Here's my suggestion for what kind of field: How about we just use a `context.Context`?
Then we can just add metadata to that context and it should even be serialized over the RPC interface automatically. So for example `metadata.AppendToOutgoingContext(ctx, "Grpc-Metadata-psbt", packet.B64Encode())`.

Then in most places the context can just be threaded through.
What do you think? Too hacky? Too much relying on assumptions on how the gRPC library works?
Not sure if the HTTP maximum header size would be an issue though... Do we expect the PSBT to be larger than 16k very often?

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@aakselrod, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
Closing due to inactivity

lightninglabs-deploy commented:
Closing due to inactivity
